docs/proposals/respect-rbac-for-resource-exclusions.md

```diff
@@ -0,0 +1,73 @@
+---
+title: Respect RBAC for Resource Inclusions/Exclusions
+
+authors:
+- "@gdsoumya"
+- "@alexmt"
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2023-05-03
+
+---
+
+# Enhancement Idea
+
+This is a proposal to provide the ability to configure argocd controller, to respect the current RBAC permissions 
+when handling resources besides the already existing resource inclusions and exclusions.
+
+## Summary
+
+Argo CD administrator will be able to configure in `argocd-cm`, whether to enable or disable(default) the feature where the controller will 
+only monitor resources that the current service account allows it to read.
+
+## Motivation
+
+Some users restrict the access of the argocd to specific resources using rbac and this feature will enable them to continue 
+using argocd without having to manually configure resource exclusions for all the resources that they don't want argocd to be managing.
+
+## Proposal 
+
+The configuration for this will be present in the `argocd-cm`, we will add new boolean field `resource.respectRBAC` in the
+cm which can be set to `true` to enable this feature, by default the feature is disabled.
+
+For the implementation there are 3 proposals :
+
+1. Modify `gitops-engine` pkg to make a `SelfSubjectAccessReview` request before adding any resource to the watch list, in this approach we are making an extra
+   api server call to check if controller has access to the resource, this does increase the no. of kubeapi calls made but is more accurate.
+2. Modify `gitops-engine` pkg to check for forbidden/unauthorized errors when listing for resources, this is more efficient approach as the
+   no. of kubeapi calls made does not change, but there is a chance of false positives as similar errors can be returned from kubeapi server or env specific proxies in other situations
+3. Combine approaches 1 and 2, in this controller will check the api response for the list call, and if it receives forbidden/unauthorized it will make the `SelfSubjectAccessReview` call.
+   This approach is accurate and at the same time, only makes extra api calls if the list calls fail in the first place.
+
+In all solutions, once controller determines that it does not have access to the resource it will stop monitoring it.
+
+### Implementation decision
+
+It was decided that we will go with approach 3 from the above list, but we shall provide 2 boolean configurations options to users :
+   - `resource.respectRBAC.strict` : This will perform both the checks i.e. whether the list call response is forbidden/unauthorized and if it is make the `SelfSubjectAccessReview` call to confirm.
+   - `resource.respectRBAC.normal` : This will only check whether the list call response is forbidden/unauthorized and skip `SelfSubjectAccessReview` call.
+
+NOTE: `strict` has higher priority so irrespective of the status of `normal` if strict is set to true strict mode will be used.
```

Comment:
It's more of a nit, but could we maybe - instead of two boolean options - have one option that takes three values: 

* `normal` (or maybe `lax`, because this is more the opposite of `strict`), 
* `strict` 
* `false` (or `disabled`, or `off`), being also the default if the field is not set.

Or as an alternative: Have two boolean values, `resource.respectRBAC.enabled` and `resource.respectRBAC.strict`, but in order to turn on `strict`, `enabled` has to be set to true as well.

I believe that would be less confusing and more ergonomic to users configuring Argo CD. But as I said, it's a nit, and I'm interested in other people's opinion.

Reply:
Oh that's a good idea I think having a single field with 3 values is fine, I would probably make it just 2 and use empty "" or missing field as disabled. Wdyt?